Codechange: Use template traits to define BaseSets.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Declarations/definitions for graphics, sound and music basesets are contained in a single file. When using these definitions, macros are used to define the set type, and macros are used to instantiate the methods.

Some BaseSet settings are set as part of the template parameters, which need to be repeated for each definition.

For different implementation details, there is specialisation but done in an odd way.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

* Split the definitions into separate files. This reduces scope slightly (though not that much because often all three are needed.)
* Use template traits to set the BaseSet parameters -- number of files and whether to search in tars.
* Use template trait to set the BaseSet type name, instead of a macro.
* Regular template specialisation for per-baseset specifics.
* Use explicit instantiation of the class, instead of per-method instantiation with a macro.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 commented on src/base_media_func.h:
```diff
@@ -38,12 +38,12 @@ extern void CheckExternalFiles();
 	const IniGroup *metadata = ini.GetGroup("metadata");
 	if (metadata == nullptr) {
-		Debug(grf, 0, "Base " SET_TYPE "set detail loading: metadata missing.");
+		Debug(grf, 0, "Base {}set detail loading: metadata missing.", SET_TYPE);
```
```suggestion
		Debug(grf, 0, "Base {}set detail loading: metadata missing.", BaseSet<Tbase_set>::SET_TYPE);
```

rubidium42 commented on src/base_media_func.h:
```diff
@@ -80,12 +80,12 @@ bool BaseSet<T, Tnum_files, Tsearch_in_tars>::FillSetDetails(const IniFile &ini,
 	const IniGroup *md5s   = ini.GetGroup("md5s");
 	const IniGroup *origin = ini.GetGroup("origin");
-	for (uint i = 0; i < Tnum_files; i++) {
+	for (uint i = 0; i < NUM_FILES; i++) {
```
```suggestion
	for (uint i = 0; i < BaseSet<Tbase_set>::NUM_FILES; i++) {
```

rubidium42 commented on src/gfxinit.cpp:
```diff
@@ -497,21 +485,21 @@ template <class Tbase_set>
 	};
 
-	for (const Tbase_set *c = BaseMedia<Tbase_set>::available_sets; c != nullptr; c = c->next) {
+	for (const GraphicsSet *c = available_sets; c != nullptr; c = c->next) {
```
```suggestion
	for (const GraphicsSet *c = BaseMedia<GraphicsSet>::available_sets; c != nullptr; c = c->next) {
```

rubidium42 commented on src/base_media_base.h:
```diff
@@ -83,7 +86,7 @@ struct BaseSet {
 	int GetNumMissing() const
 	{
-		return Tnum_files - this->found_files;
+		return NUM_FILES - this->found_files;
```
```suggestion
		return BaseSet<T>::NUM_FILES - this->found_files;
```

rubidium42 commented on src/base_media_base.h:
```diff
@@ -93,7 +96,7 @@ struct BaseSet {
 	int GetNumInvalid() const
 	{
-		return Tnum_files - this->valid_files;
+		return NUM_FILES - this->valid_files;
```
```suggestion
		return BaseSet<T>::NUM_FILES - this->valid_files;
```

rubidium42 commented on src/base_media_func.h:
```diff
@@ -25,7 +25,7 @@ extern void CheckExternalFiles();
 	item = metadata->GetItem(name); \
 	if (item == nullptr || !item->value.has_value() || item->value->empty()) { \
-		Debug(grf, 0, "Base " SET_TYPE "set detail loading: {} field missing.", name); \
+		Debug(grf, 0, "Base {}set detail loading: {} field missing.", SET_TYPE, name); \
```
```suggestion
		Debug(grf, 0, "Base {}set detail loading: {} field missing.", BaseSet<T>::SET_TYPE, name); \
```

rubidium42 commented on src/base_media_func.h:
```diff
@@ -80,12 +80,12 @@ bool BaseSet<T, Tnum_files, Tsearch_in_tars>::FillSetDetails(const IniFile &ini,
+		item = files != nullptr ? files->GetItem(BaseSet<T>::file_names[i]) : nullptr;
 		if (item == nullptr || (!item->value.has_value() && !allow_empty_filename)) {
-			Debug(grf, 0, "No " SET_TYPE " file for: {} (in {})", BaseSet<T, Tnum_files, Tsearch_in_tars>::file_names[i], full_filename);
+			Debug(grf, 0, "No {} file for: {} (in {})", SET_TYPE, BaseSet<T>::file_names[i], full_filename);
```
```suggestion
			Debug(grf, 0, "No {} file for: {} (in {})", BaseSet<T>::SET_TYPE, BaseSet<T>::file_names[i], full_filename);
```

PeterN committed:
Codechange: Split baseset type definitions into separate files.

PeterN committed:
Codechange: Use traits to define BaseSet parameters.

PeterN committed:
Codechange: Use template specialisation and instantiation for BaseMedia methods.

Specialisations seem to be the correct way to specialise, rather than redefining the base template.

This removes a macro which instantiated methods individually.
